DRILL-8235: Add Storage Plugin for Google Sheets

cgivre opened the pull request:
# [DRILL-8235](https://issues.apache.org/jira/browse/DRILL-8235): Add Storage Plugin for Google Sheets

## Description
Google Sheets is a very commonly used data source among business users.  Presto and other query engines do include integrations with Google Sheets and so it would be useful for Drill to add this functionality. 

The proposed plugin supports both reading and writing to Google Sheets. 

Here's a quick demo: 
https://www.youtube.com/watch?v=pS2FKxrluBA

This PR makes a few addition minor modifications that are not directly related to Google Sheets.

* Modification to the OAuth token storage to allow plugins besides HTTP to use Drill's token storage mechanism
* Modification to the OAuth token storage to store the token expiration time.
* Reapplication of `.gitignore` and removing some files which should have been ignored.  Note this caused a unit test to break, so I had to rename a file to fix this.
* This PR also integrates [DRILL-8270](https://issues.apache.org/jira/browse/DRILL-8270)

## Documentation
Docs in README.

- [x] Finalize Unit Tests
- [x] Fix writer
- [x] Remove logback file.
- [x] Code cleanup
- [x] Fix LGTM bugs


## Testing
Manually tested and added unit tests.

jnturton commented:
Nice demo vid! Before I dive in, I see that a review has been requested but the PR is also marked draft. Can I switch it over to reviewable? Btw, I like the sheet specification syntax here and it's something I wondered about when doing a bug fix on the Excel format plugin recently. But I guess Excel being a format plugin, this syntax is not available and we have to stick with the format config opt that we have?

cgivre commented:
> 

Thanks @jnturton !!  You certainly may start reviewing if you like.  I left it as draft for the moment as there is still some minor cleanup and debugging that I have to do.  I don't anticipate any major changes, but I'm adding more unit tests and fixing all the LGTM / CodeQL issues and stuff like that.  If you want to start... please feel free.

cgivre commented:
@jnturton This is now ready for review.

jnturton commented on contrib/storage-googlesheets/src/test/java/org/apache/drill/exec/store/googlesheets/TestGoogleSheetsQueries.java:
```diff
@@ -0,0 +1,317 @@
+  }
+
+  @Test
+  public void testProjectPushdown() throws Exception {
```
Isn't it necessary to look in the plan string for the projected columns to test the pushdown part? I would expect this test to pass even if there was no projection pushdown...

cgivre replied:
Fixed.

jnturton commented on contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/GoogleSheetsBatchReader.java:
```diff
@@ -0,0 +1,301 @@
+import org.apache.drill.common.exceptions.UserException;
+import org.apache.drill.common.expression.SchemaPath;
+import org.apache.drill.common.types.TypeProtos.MinorType;
+import org.apache.drill.exec.physical.impl.scan.framework.ManagedReader;
```
```suggestion
import org.apache.drill.exec.physical.impl.scan.v3.ManagedReader;
```
Can we use the EVF2 API here?

cgivre replied:
@jnturton It turns out that EVF2 is missing a few bits to fully convert storage plugins.  There isn't a ManagedReader class for non-file systems.  I'd be happy to do that, but if it's ok with you, could we make that a separate PR?

jnturton replied:
> There isn't a ManagedReader class for non-file systems. I'd be happy to do that, but if it's ok with you, could we make that a separate PR?

Sounds good, thanks.

cgivre commented on contrib/native/client/patches/zookeeper-3.4.6-x64.patch:
```diff
@@ -1,163 +0,0 @@
-From 64697ddd8a90f29d1693658f04e975e435e3c869 Mon Sep 17 00:00:00 2001
```
I don't know.  Patch files are supposed to be excluded by the .gitignore.  What happened was I reapplied Drill's `.gitignore` and it deleted that patch file.   I would assume this file should not be included.

jnturton replied:
It's very old:
```
commit d944918ff8a1f76d9b4c400512a130bff7fb2c5f
Author: Parth Chandra <pchandra@maprtech.com>
Date:   Wed Jun 11 18:10:34 2014 -0700

    DRILL-1021: Windows build
```
@vdiravka any thoughts about this patch file?

jnturton replied:
If this PR really wants to be the one to delete this file from version control then I think it should include some comment  in the description like "this patch was added for X and is no longer relevant". I can see that it's a C code patch to ZooKeeper with a title of `[PATCH] Allow zookeeper to build in x64` which makes me 99% sure it's completely obsolete. @vdiravka, @vvysotskyi any advice for us here?

vdiravka replied:
There is [readme.win.txt](https://github.com/apache/drill/blob/master/contrib/native/client/readme.win.txt#L135) about it:
> The 3.4.6 release of Zookeeper does not build correctly on 64 bit windows. To fix that for the 64 bit build, apply patch zookeeper-3.4.6-x64.patch

cgivre replied:
@vdiravka Forgive my lack of knowledge here, but the current `.gitignore` excludes patch files.  Is there a way to ignore all patch files, but leave that one?  How would you suggest we fix this?

@jnturton Aside from this patch file, are we good to go, pending CI?

jnturton replied:
I think the .gitignore is working properly already because all of our previous commits have left zookeeper-3.4.6-x64.patch in peace. How about we do it this way? If this PR wants to delete the patch then it should simultaneously update readme.win.txt loosely saying "ZooKeeper 3.4.6 is ancient, ZooKeeper 3.5.7 is current and does not need any patch to build on Windows so the patch remains in the Git history only". Alternatively, we revert the deletion here and carry the cruft a bit longer.

+1 one from me otherwise.

cgivre replied:
Sounds good.  Updating README.

cgivre replied:
Done!

vdiravka replied:
@cgivre In any case, if the issue persists, there is a [way](https://stackoverflow.com/questions/987142/make-gitignore-ignore-everything-except-a-few-files) how to exclude only one file from the pattern:
`!contrib/native/client/patches/zookeeper-3.4.6-x64.patch`

cgivre commented on .gitignore:
```diff
@@ -28,3 +28,6 @@ exec/jdbc-all/dependency-reduced-pom.xml
 venv/
 tools/venv/
+
+# Directory to store oauth tokens for testing Googlesheets Storage plugin
```
Fixed.

cgivre commented on contrib/storage-googlesheets/README.md:
````diff
@@ -0,0 +1,156 @@
+}
+```
+
+With the exception of the clientID, client_secret and redirects, you should not have to modify any of the other parameters in the configuration. 
````
Fixed

cgivre commented on contrib/storage-googlesheets/README.md:
```diff
@@ -0,0 +1,156 @@
+* `extractHeaders`:  When `true`, Drill will treat the first row of your data as headers.  When `false` Drill will assign column names like `field_n` for each column.
+
+### Authenticating with Google
+Once you have configured Drill to query GoogleSheets, there is one final step before you can access data.  You must authenticate the application (Drill) with GoogleSheets.  After you have saved your GoogleSheets configuration, navigate back to the configuration screen for your plugin and click on `Authorize`. A new window should appear which will prompt you to authenticate with Google services.   ...
```
Fixed here and elsewhere.

cgivre commented on contrib/storage-googlesheets/README.md:
````diff
@@ -0,0 +1,156 @@
+```
+Note that you must specify the tab name to successfully query GoogleSheets.
+
+### Using Aliases
````
Thx!!

cgivre commented on contrib/storage-googlesheets/README.md:
```diff
@@ -0,0 +1,156 @@
+supply a schema or other information to allow Drill to identify the data types of a column.  At present, here is how Drill will map your data:
+* Numbers:  All numeric columns will be mapped to `DOUBLE` data types
+* Boolean:  Columns containing `true/false` will be mapped to the `BOOLEAN` type
+* Time, Date, Timestamp:  Temporal fields will be mapped to the correct type.  You can disable able temporal fields by setting the config option `XXX` to `false`.
```
Removed... I was going to do this, but I think `allTextMode` will suffice.

cgivre commented on contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/GoogleSheetsPushDownListener.java:
```diff
@@ -0,0 +1,149 @@
+
+    @Override
+    public ExprNode accept(ExprNode node) {
+      if (node instanceof OrNode) {
+        return null;
+      } else if (node instanceof ColRelOpConstNode) {
+        return null;
+        //return acceptRelOp((ColRelOpConstNode) node);
+      } else {
+        return null;
+      }
```
For now yes.  If we can figure out how to use the Google SDK for filter pushdowns, we should uncomment out that line.  I'll add a comment there, explaining.  I'll do my best to keep the snark factor low, but no promises.

cgivre commented on contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/GoogleSheetsStoragePlugin.java:
```diff
@@ -0,0 +1,219 @@
+    // existing plugins perform filter push-down at the physical
+    // phase, which also works fine if push-down is independent of
+    // parallelization.
+    if (FilterPushDownUtils.isFilterPushDownPhase(phase) || phase == PlannerPhase.LOGICAL) {
```
Fixed.

cgivre commented on contrib/storage-googlesheets/src/test/java/org/apache/drill/exec/store/googlesheets/TestGoogleSheetsLimitPushdown.java:
```diff
@@ -0,0 +1,171 @@
+
+import static org.junit.Assert.fail;
+
+@Ignore("This test requires a live connection to GoogleSheets.  Please run tests manually.")
```
Fixed, here and elsewhere.

cgivre commented on contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/GoogleSheetsGroupScan.java:
```diff
@@ -0,0 +1,369 @@
+    }
+
+    // No good estimates at all, just make up something.
+    double estRowCount = 10_000;
```
I'll add a TODO.  I think you are correct that we can get the max row count, but I'll save that for future work.

cgivre replied:
Added TODO.

cgivre commented on contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/utils/GoogleSheetsTypifier.java:
```diff
@@ -0,0 +1,346 @@
+ * This class attempts to infer the data type of an unknown data type. It is somewhat
+ * configurable.  This was sourced from <a href="https://gist.github.com/awwsmm/56b8164410c89c719ebfca7b3d85870b">this code on github</a>.
+ */
+public class GoogleSheetsTypifier {
```
Moved to `common`.

cgivre commented on contrib/storage-googlesheets/README.md:
```diff
@@ -0,0 +1,156 @@
+At the time of writing, we have not yet documented `USER_TRANSLATION` fully, however we will update this readme once that is complete.
+
+## Querying Data
+Once you have configured Drill to query 
```
Fixed

cgivre commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/security/OAuthTokenCredentials.java:
```diff
@@ -0,0 +1,82 @@
+
+import java.util.Map;
+
+public class OAuthTokenCredentials {
```
Good catch!  I had been working on this for some time, and I think I created that before we merged the final version of OAuth support.  In any event, the duplicate class has been removed.

cgivre commented on contrib/storage-googlesheets/src/main/java/org/apache/drill/exec/store/googlesheets/columns/GoogleSheetsDateColumnWriter.java:
```diff
@@ -0,0 +1,45 @@
+
+import java.time.LocalDate;
+
+public class GoogleSheetsDateColumnWriter extends GoogleSheetsColumnWriter {
```
Done!

cgivre commented on contrib/storage-googlesheets/src/test/java/org/apache/drill/exec/store/googlesheets/TestGoogleSheetsWriter.java:
```diff
@@ -0,0 +1,135 @@
+
+    /*String query = "CREATE TABLE googlesheets.`test_sheet`.`test_table` (ID, NAME) AS " +
+      "SELECT * FROM (VALUES(1,2), (3,4))";*/
+   String sql = "SELECT * FROM table(cp.`data/Drill_Test_Data.xlsx` (type => 'excel', sheetName => 'MixedSheet'))";
```
I removed this and added a comment about it.  With the current implementation, when you execute a CTAS query, you do so with the desired file name. IE:

```sql
CREATE TABLE googlesheets.myfile AS
SELECT <stuff>
FROM <data>
```

The problem occurs when you go to test that and you need the file ID.  Google's SDK doesn't give you a way that I found to go from file name to file ID.

cgivre commented on logical/src/main/java/org/apache/drill/common/logical/StoragePluginConfig.java:
```diff
@@ -33,6 +34,7 @@
 
 @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
 @JsonInclude(JsonInclude.Include.NON_DEFAULT)
+@JsonFormat(with = JsonFormat.Feature.ACCEPT_CASE_INSENSITIVE_PROPERTIES)
```
I can add this to the generic FormatPlugin super class.  Should that be a separate PR?

jnturton replied:
Yes, thank you, let's make case insensitive properties its own PR. Please revert it here.

cgivre replied:
Done!

cgivre commented:
@jnturton Thank you very much for your review of this beast.  I believe I have addressed all your comments, as well as fixed the final bugs with the writer.

jnturton reviewed (changes requested):
Thanks for all the fixes. Just two minor items remaining from me.
